infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java

```diff
@@ -139,6 +132,66 @@ protected void internalStart(Map<String, String> startOptions) throws Infrastruc
     }
   }
 
+  /**
+   * Bootstraps machines in parallel.
+   *
+   * @throws InfrastructureException when any error occurs while bootstrapping machines
+   * @throws InterruptedException when machines bootstrapping was interrupted
+   */
+  private void bootstrapMachines() throws InfrastructureException, InterruptedException {
+    List<CompletableFuture> checkMachines = new ArrayList<>(machines.size());
+    List<BootstrapMachineTask> tasks = new ArrayList<>(machines.size());
+
+    for (OpenShiftMachine machine : machines.values()) {
+      BootstrapMachineTask task = new BootstrapMachineTask(machine);
+      tasks.add(task);
+      checkMachines.add(CompletableFuture.runAsync(task, pool.getExecutor()));
+    }
+    try {
+      CompletableFuture.allOf(checkMachines.toArray(new CompletableFuture[0])).join();
+    } catch (CompletionException e) {
+      tasks.forEach(BootstrapMachineTask::interrupt);
+      if (e.getCause() instanceof InterruptedException) {
+        throw (InterruptedException) e.getCause();
+      } else if (e.getCause() instanceof InfrastructureException) {
+        throw (InfrastructureException) e.getCause();
+      } else {
+        throw new InfrastructureException(e.getCause().getMessage(), e.getCause());
```

Comment:
I would say that in this situation InternalRuntimeException is better. @mshaposhnik @sleshchenko WDYT?

Reply:
Didi you mean InternalInfrastructureException?

Reply:
Yes, sorry for confusing.